s2a/src/main/java/io/grpc/s2a/channel/S2AGrpcChannelPool.java

```diff
@@ -0,0 +1,109 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.s2a.channel;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import io.grpc.Channel;
+import io.grpc.internal.ObjectPool;
+import javax.annotation.concurrent.ThreadSafe;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Manages a gRPC channel pool and a cached gRPC channel to be used for communication with the S2A.
+ */
+@ThreadSafe
+public final class S2AGrpcChannelPool implements S2AChannelPool {
+  private static final int MAX_NUMBER_USERS_OF_CACHED_CHANNEL = 100000;
+  private final ObjectPool<Channel> channelPool;
+
+  @GuardedBy("this")
+  private @Nullable Channel cachedChannel;
+
+  @GuardedBy("this")
+  private int numberOfUsersOfCachedChannel = 0;
+
+  private enum State {
+    OPEN,
+    CLOSED,
+  }
+  
+  ;
+
+  @GuardedBy("this")
+  private State state = State.OPEN;
+
+  public static S2AChannelPool create(ObjectPool<Channel> channelPool) {
+    checkNotNull(channelPool, "Channel pool should not be null.");
+    return new S2AGrpcChannelPool(channelPool);
+  }
+
+  private S2AGrpcChannelPool(ObjectPool<Channel> channelPool) {
+    this.channelPool = channelPool;
+  }
+
+  /**
+   * Retrieves a channel from {@code channelPool} if {@code channel} is null, and returns {@code
+   * channel} otherwise.
+   *
+   * @return a {@link Channel} obtained from the channel pool.
+   */
+  @Override
+  public synchronized Channel getChannel() {
+    checkState(state.equals(State.OPEN), "Channel pool is not open.");
+    checkState(
+        numberOfUsersOfCachedChannel < MAX_NUMBER_USERS_OF_CACHED_CHANNEL,
+        "Max number of channels have been retrieved from the channel pool.");
+    if (cachedChannel == null) {
+      cachedChannel = channelPool.getObject();
+    }
+    numberOfUsersOfCachedChannel += 1;
+    return cachedChannel;
+  }
+
+  /**
+   * Returns {@code channel} to {@code channelPool}.
+   *
+   * <p>The caller must ensure that {@code channel} was retrieved from this channel pool.
+   */
+  @Override
+  public synchronized void returnToPool(Channel channel) {
+    checkState(state.equals(State.OPEN), "Channel pool is not open.");
+    checkArgument(
+        cachedChannel != null && numberOfUsersOfCachedChannel > 0 && cachedChannel.equals(channel),
+        "Cannot return the channel to channel pool because the channel was not obtained from"
+            + " channel pool.");
+    numberOfUsersOfCachedChannel -= 1;
+    if (numberOfUsersOfCachedChannel == 0) {
+      channelPool.returnObject(channel);
+      cachedChannel = null;
+    }
+  }
+
+  @Override
+  public synchronized void close() {
```

Comment:
Deleted the Class in https://github.com/grpc/grpc-java/pull/11584/commits/f7574613753dea7ae4c178cc33336a858c0571c0. The other thing this class provides is an upper limit on the allowed users of the channel(`MAX_NUMBER_USERS_OF_CACHED_CHANNEL`); does gRPC provide an upper limit on the number of users per channel, so that it’s ok we no longer have this check?

Reply:
Ok, I'll add that the s2a server grpc layer